### PR TITLE
JSON: remove dynamicProperty and dynamicType if empty; add _vimtype to DataObjects

### PR DESCRIPTION
**Pull request by jeking3**
Remove dynamicProperty and dynamicType if empty during JSON encode; add _vimtype to DataObjects

Follow-on to https://github.com/vmware/pyvmomi/pull/732

Improvements:

When `dynamicProperty` and `dynamicType` are empty, they get stripped from the results.  Additionally every DataObject now also gets a `_vimtype` property which allows one to query the json by type of device.  An example structure coming out is the list of datastores in the vm.config:
```
        "datastoreUrl": [
            {
                "_vimtype": "vim.vm.ConfigInfo.DatastoreUrlPair",
                "name": "datastore1",
                "url": "/vmfs/volumes/5b7ee5c9-4e8f1921-93b3-90b11c51ac73"
            }
        ],
```

Adding the type is important.  Before if you wanted to query for devices based on type you had to inspect the `deviceInfo.summary` which is brittle.  Now you can search for `_vimtype` in the devices.
```
                {
                    "_vimtype": "vim.vm.device.VirtualLsiLogicSASController",
                    "backing": null,
                    "busNumber": 0,
                    "connectable": null,
                    "controllerKey": 100,
                    "device": [
                        2000,
                        2001,
                        2002
                    ],
                    "deviceInfo": {
                        "_vimtype": "vim.Description",
                        "label": "SCSI controller 0",
                        "summary": "LSI Logic SAS"
                    },
                    "hotAddRemove": true,
                    "key": 1000,
                    "scsiCtlrUnitNumber": 7,
                    "sharedBus": "noSharing",
                    "slotInfo": {
                        "_vimtype": "vim.vm.device.VirtualDevice.PciBusSlotInfo",
                        "pciSlotNumber": 160
                    },
                    "unitNumber": 3
                },
                ...
```

### CODE REVIEW

**Comment by jeking3** [comment]
@pgbidkar any chance we could also get this merged?

File: pyVmomi/VmomiSupport.py [hunk 1]
```diff
@@ -316,6 +322,13 @@ def __init__(self, *args, **kwargs):
         self._explode = kwargs.pop('explode', None)
         super(VmomiJSONEncoder, self).__init__(*args, **kwargs)
 
+    def _clean_dynamic(self, result):
+        if 'dynamicProperty' in result and len(result['dynamicProperty']) == 0:
```
Comment (tianhao64) [inline pyVmomi/VmomiSupport.py]: Why do you check len here but use 'not.. ' below?

**Comment by jeking3** [comment]
Any chance we could get a 6.7.1.2018.12 release with this included?  I have a number of upstream pull requests on other projects (Ansible, StackStorm) waiting for this code to make it into a release.

**Comment by tianhao64** [comment]
@jeking3 Could you comment on my previous question on why removing these two properties are necessary?

**Comment by jeking3** [comment]
I'm working on this now, and given your concern about it, I will make the default `False` for this feature and add an integration test.

**Comment by tianhao64** [comment]
@jeking3 Thanks for the change!

**Comment by jeking3** [comment]
Thanks - could you let me know here or privately what the next release date will be?  This will allow me to unblock a number of pull requests in various projects.

**Comment by pgbidkar** [comment]
Hi @jeking3 Thanks for Contributing. 
We are discussing regarding releasing pyVmomi with latest merges. We will keep you updated.

Thanks

**Pull request merged**